### ISSUE TITLE
a question about delta_yolo_box

### ISSUE THREAD

**Issue by GYC1996** [AlexeyAB/darknet#2733]
Hi, AlexeyAB:
    Now, I have a question.
    In paper, the box use  sum of squared error loss to calculate. And the x,y use sigmoid to activate but the w,h not. So I think the delta[x], delta[y] should be different to delta[w],delta[h]. But in code

    float tx = (truth.x*lw - i);
    float ty = (truth.y*lh - j);
    float tw = log(truth.w*w / biases[2*n]);
    float th = log(truth.h*h / biases[2*n + 1]);

    delta[index + 0*stride] = scale * (tx - x[index + 0*stride]);
    delta[index + 1*stride] = scale * (ty - x[index + 1*stride]);
    delta[index + 2*stride] = scale * (tw - x[index + 2*stride]);
    delta[index + 3*stride] = scale * (th - x[index + 3*stride]);
 
they are both the same expression. I don't konw why.

**Comment by AlexeyAB** [comment]
@GYC1996 Hi,

They are not the same, just substitute the value of variables `tw` into delta formula.

Sigmoid for x,y are calculated here: https://github.com/AlexeyAB/darknet/blob/f4825906916130ac1c32eb69c9fb812aacde1838/src/yolo_layer.c#L449

**Comment by GYC1996** [comment]
> @GYC1996 Hi,
> 
> They are not the same, just substitute the value of variables `tw` into delta formula.
> 
> Sigmoid for x,y are calculated here:
> 
> [darknet/src/yolo_layer.c](https://github.com/AlexeyAB/darknet/blob/f4825906916130ac1c32eb69c9fb812aacde1838/src/yolo_layer.c#L449)
> 
> Line 449 in [f482590](/AlexeyAB/darknet/commit/f4825906916130ac1c32eb69c9fb812aacde1838)
> 
>  activate_array_ongpu(l.output_gpu + index, 2*l.w*l.h, LOGISTIC);    // x,y

yes, I know. but the gradient of x,y maybe include of sigmoid?

**Comment by WangDeyu** [comment]
The same problem as @GYC1996

**Comment by GYC1996** [comment]
> The same problem as @GYC1996

I have added "p * (1-p)" in the expression, the map getting more stable with iteration, but need more time to training. The max value of map just a little improvement, so I think it maybe the author remove on purpose.

**Comment by WangDeyu** [comment]
> > The same problem as @GYC1996
> 
> I have added "p * (1-p)" in the expression, the map getting more stable with iteration, but need more time to training. The max value of map just a little improvement, so I think it maybe the author remove on purpose.

Maybe you are right! Both p and (1-p) are in range of 0 to 1, they do not change the direction of gradient so as to update in right direction, but just like scale factor to control the gradient. @GYC1996 @AlexeyAB

**Comment by cnLevi** [comment]
@WangDeyu @GYC1996 
Hey guys, check this out: https://peterroelants.github.io/posts/cross-entropy-logistic/
I think it's because of the gradient of the cross-entropy loss function with sigmoid as activate function is y-y_hat...

**Comment by cnLevi** [comment]
#1695

**Comment by AlexeyAB** [comment]
@GYC1996 

> I have added "p * (1-p)" in the expression, the map getting more stable with iteration, but need more time to training. The max value of map just a little improvement, so I think it maybe the author remove on purpose.

Can you show your changed code?

**Comment by GYC1996** [comment]
> @GYC1996
> 
> I have added "p * (1-p)" in the expression, the map getting more stable with iteration, but need more time to training. The max value of map just a little improvement, so I think it maybe the author remove on purpose.
> 
> Can you show your changed code?

    delta[index + 0*stride] = scale * (tx - x[index + 0*stride]) * (1 - x[index + 0*stride]) * x[index + 0*stride];
    delta[index + 1*stride] = scale * (ty - x[index + 1*stride])* (1 - x[index + 1*stride]) * x[index + 1*stride];

**Comment by AlexeyAB** [comment]
@GYC1996 Why do you think that you should use `p * (1-p)` ?

**Comment by WangDeyu** [comment]
@AlexeyAB loss = (tx - x[index + 0*stride])^2 = scale(tx - sigmoid(px))^2 
                    The gradient of px =  scale(tx - sigmoid(px)) * (1 - sigmoid(px)) * sigmoid(px)
                    I know there is sigmoid activation in yolo_forward, but it dosen't have derivative of sigmoid in backward, if we want to get the delta of px, we should use p*(1-p)

**Comment by GYC1996** [comment]
> @AlexeyAB
> loss = (tx - x[index + 0*stride])^2 = scale(tx - sigmoid(px))^2 
> The gradient of px =  scale(tx - sigmoid(px)) * (1 - sigmoid(px)) * sigmoid(px)
> I know there is sigmoid activation in yolo_forward, but it dosen't have derivative of sigmoid in backward, if we want to get the delta of px, we should use p*(1-p)

yes

**Comment by lz20061213** [comment]
i think you are right

**Comment by wuhy08** [comment]
@AlexeyAB  @GYC1996  @WangDeyu 
I am investigating the loss function of YOLOv3 and come across this thread. I think there are some contradictions between the paper and the code.
In Redmon's YOLOv3 paper, the center coordinate of the bounding box (_b<sub>x</sub>_, _b<sub>y</sub>_) was calculated the following way:
_b<sub>x</sub>=&sigma;(t<sub>x</sub>)+c<sub>x</sub>_,
_b<sub>y</sub>=&sigma;(t<sub>y</sub>)+c<sub>y</sub>_.
Where _t<sub>x</sub>_ and _t<sub>y</sub>_ are the output from the previous layer without any activation, _c<sub>x</sub>_ and _c<sub>y</sub>_ are the top left corner.
And in paper, the sum of squared error is used to calculate the loss function:
_L=(t<sub>x</sub> - t'<sub>x</sub>)<sup>2</sup>+(t<sub>y</sub> - t'<sub>y</sub>)<sup>2</sup>+..._ , where  _t'<sub>x</sub>_ and _t'<sub>y</sub>_ are the coordinate offset value of the ground truth from the following calculation:
_t'<sub>x</sub>=&sigma; <sup>-1</sup>(b'<sub>x</sub>-c<sub>x</sub>)_,
where _b'<sub>x</sub>_ is the raw coordinate of the ground truth.

However, in the code (both original and @AlexeyAB 's), instead of using _t'<sub>x</sub>_ and _t<sub>x</sub>_ , _&sigma;(t<sub>x</sub>)_ and  _&sigma;(t'<sub>x</sub>)_ are used. In the code, the loss function is not calculated, only the gradient of the loss function, and it was:
_dL/dt<sub>x</sub>=&sigma;(t'<sub>x</sub>) - &sigma;(t<sub>x</sub>)_ .
This is equivalent to the gradient of Binary Cross Entropy (BCE) loss function.

As a conclusion, the code implements the BCE loss function between the ground truth (offset but not inverse sigmoid-ed) and the prediction (after sigmoid activation), not the MSE loss function as described in the paper.

If we want to replicate the paper, we firstly should not pass the prediction through the activation:
https://github.com/AlexeyAB/darknet/blob/08e8e0c8c2d3cddfb6ee065e813ae5023f9568d2/src/yolo_layer.c#L405
and we should do an inverse operation of the ground truth:
https://github.com/AlexeyAB/darknet/blob/08e8e0c8c2d3cddfb6ee065e813ae5023f9568d2/src/yolo_layer.c#L99-L100

But which loss function gives better result is to be debated.

**Comment by AlexeyAB** [comment]
@wuhy08 Hi,


> If we want to replicate the paper, we firstly should not pass the prediction through the activation:
> https://github.com/AlexeyAB/darknet/blob/08e8e0c8c2d3cddfb6ee065e813ae5023f9568d2/src/yolo_layer.c#L405
> and we should do an **inverse operation** of the ground truth:
> https://github.com/AlexeyAB/darknet/blob/08e8e0c8c2d3cddfb6ee065e813ae5023f9568d2/src/yolo_layer.c#L99-L100
> 
> But which loss function gives better result is to be debated.

What do you mean by inverse operation?

Do you mean that we should apply LOGISTIC for tx and ty only for Detection, while for training we should use sourсe tx, ty and inversed ground truth?
_b<sub>x</sub>=&sigma;(t<sub>x</sub>)+c<sub>x</sub>_,
_b<sub>y</sub>=&sigma;(t<sub>y</sub>)+c<sub>y</sub>_.

**Comment by wuhy08** [comment]
Hi @AlexeyAB 
By saying "inverse operation", I meant inversion sigmoid function: _&sigma; <sup>-1</sup>(x)=log(x/(1-x))_ .

For your second question, I mean that if we want to match the code with the paper, we should use tx, ty and inversed ground truth for training. And for testing (inference), we apply sigmoid (logistic) function to tx, ty and then add the offset value.

I am not sure which will give a better result.

Since you are here, I have another question: 

It seems that in YOLOv3's loss function, there aren't hyperparameters to scale different parts of the loss function (coordinate loss, class loss, objectness loss) except for the scale in `delta_yolo_box` function, is that correct? Of course, for MSE losses (only wh_loss), there is an implicit factor of 0.5, since the derivative of MSE loss should have a factor of 2,  which is missing in the code.

Therefore, `total_loss = delta_yolo_box_scale*(xy_loss+0.5*wh_loss) + class_loss + objectness_loss`. Is that correct?

Thank you!

**Comment by AlexeyAB** [comment]
@wuhy08 
Do you mean that we get `2` from derivatives of square (from derivatives of MSE)?
And since we mix the losses of Binary Cross Entropy (BCE) `dL/dtx=σ(t'x) - σ(tx)` and Mean Squared Error (MSE) `dL/dtx=(2/n)*sum((t'x) - (tx))` , then we should use scaling `1/2` for MSE-loss?

Is it about this? (multiplier 2 is burned) https://datascience.stackexchange.com/a/52159/37736
http://mccormickml.com/2014/03/04/gradient-descent-derivation/

**Comment by lz20061213** [comment]
> @AlexeyAB @GYC1996 @WangDeyu
> I am investigating the loss function of YOLOv3 and come across this thread. I think there are some contradictions between the paper and the code.
> In Redmon's YOLOv3 paper, the center coordinate of the bounding box (_bx_, _by_) was calculated the following way:
> _bx=σ(tx)+cx_,
> _by=σ(ty)+cy_.
> Where _tx_ and _ty_ are the output from the previous layer without any activation, _cx_ and _cy_ are the top left corner.
> And in paper, the sum of squared error is used to calculate the loss function:
> _L=(tx - t'x)2+(ty - t'y)2+..._ , where _t'x_ and _t'y_ are the coordinate offset value of the ground truth from the following calculation:
> _t'x=σ -1(b'x-cx)_,
> where _b'x_ is the raw coordinate of the ground truth.
> 
> However, in the code (both original and @AlexeyAB 's), instead of using _t'x_ and _tx_ , _σ(tx)_ and _σ(t'x)_ are used. In the code, the loss function is not calculated, only the gradient of the loss function, and it was:
> _dL/dtx=σ(t'x) - σ(tx)_ .
> This is equivalent to the gradient of Binary Cross Entropy (BCE) loss function.
> 
> As a conclusion, the code implements the BCE loss function between the ground truth (offset but not inverse sigmoid-ed) and the prediction (after sigmoid activation), not the MSE loss function as described in the paper.
> 
> If we want to replicate the paper, we firstly should not pass the prediction through the activation:
> https://github.com/AlexeyAB/darknet/blob/08e8e0c8c2d3cddfb6ee065e813ae5023f9568d2/src/yolo_layer.c#L405
> 
> 
> and we should do an inverse operation of the ground truth:
> https://github.com/AlexeyAB/darknet/blob/08e8e0c8c2d3cddfb6ee065e813ae5023f9568d2/src/yolo_layer.c#L99-L100
> 
> But which loss function gives better result is to be debated.

My opinion,  in the code, the σ(output[x]) is means the predict tx, so in loss function, it will be (tx - σ(output[x]))^2.  That's a little different from the paper. And in the paper, the author discuss the different between the linear x, y predictions instead of logistic, which led to a couple point drop in mAP.

**Comment by AlexeyAB** [comment]
@lz20061213 There are 2 suggestions which you can try and compare the mAP with default loss: https://github.com/AlexeyAB/darknet/issues/4594#issuecomment-569423126

**Comment by h3z** [comment]
notation is such confusing between paper and code. so forget them all.
lets say, `bx = sigmoid(😀) + cx`, and truth is 😞
**in paper:**
it's clear, loss is: `(😀 - 😞)^2`

**in code** it's, 
```
```

**Comment by h3z** [comment]
notation is such confusing between paper and code. so forget them all.
lets say, bx = sigmoid(😀) + cx, and truth is 😞
in paper:
it's clear, loss is: (😀 - 😞)^2

in code it's,
```
x1 = sigmoid(😀)
x2 = sigmoid(x1)
# same for 😞
loss = binary_crossentropy(x2, sigmoid(sigmoid(😞)))
```

and when back propagation, the gradients is same, so, maybe notation is confusing, but  the results are same.

**Comment by h3z** [comment]
FINALLY, I figured out paper and code are absolutely equivalent, just a math trick (hope I am right...)

its simple to explain
```
conv(activation='sigmoid') -> MSE
```
equivalent with
```
conv(activation='sigmoid) -> sigmoid() -> BCE
```
compute the gradient for conv  (model output layer), is exactly same. 
So you don't need change anything to replicate paper. @wuhy08 

and the reason why conv with sigmoid is explained in the paper
> Linear x, y predictions instead of logistic. We tried using a linear activation to directly predict the x, y offset instead of the logistic activation. This led to a couple point drop in mAP.

**Comment by h3z** [comment]
1. in paper said, xywh use MSE
2. loss function's final value is not the most important thing, but the gradient is the most important.Because gradient is really the key for model's training.
3. so, when paper say loss function use MSE, means, model's output layer should get the gradient `t* - t`
4. its ok for wh, just `tw* - tw`, but for xy, in the paper， they said
> We tried using a linear activation to directly predict the x, y offset instead of the logistic activation. This led to a couple point drop in mAP.
5. so this means, when output layer output the value x,y, we can't use `tx = x`, but`tx = σ(x)`
6. so, for output layer, output the xywh, correspond to MSE, we need get the gradient
```
σ(tx*) - σ(tx)
σ(ty*) - σ(ty)
tw* - tw
th* - th
```
7. how to get first two line graident? just 
```
conv(name='x') -> sigmoid -> BSE
conv(name='y') -> sigmoid -> BSE
conv(name='w') -> MSE
conv(name='h') -> MSE

note all conv without activation
```


so, the code is exactly replicate the paper @wuhy08
